File: LaoCai_BaoThang_2324/Cau2/CAU02.py
def sieve_of_eratosthenes(max_num):
    is_prime = [True] * (max_num + 1)
    p = 2
    is_prime[0] = is_prime[1] = False  # 0 và 1 không phải là số nguyên tố
    
    while (p * p <= max_num):
        if is_prime[p]:
            for i in range(p * p, max_num + 1, p):
                is_prime[i] = False
        p += 1

    return is_prime

def count_primes_in_range(A, B):
    # Tạo danh sách số nguyên tố từ 1 đến B
    is_prime = sieve_of_eratosthenes(B)
    
    # Đếm số lượng số nguyên tố trong đoạn [A, B]
    count = sum(1 for i in range(A, B + 1) if is_prime[i])
    
    return count

# Đọc dữ liệu vào từ file
with open("CAU02.INP", "r") as infile:
    A, B = map(int, infile.readline().strip().split())

# Đếm số lượng số nguyên tố
result = count_primes_in_range(A, B)

# Ghi kết quả ra file
with open("CAU02.OUT", "w") as outfile:
    outfile.write(str(result) + '\n')

File: THCS/2023_2024/LaoCai_BaoThang_2324/Cau2/CAU02.cpp
#include <iostream>
#include <vector>
#include <fstream>

using namespace std;

vector<bool> sieve_of_eratosthenes(int max_num) {
    vector<bool> is_prime(max_num + 1, true);
    is_prime[0] = is_prime[1] = false; // 0 và 1 không phải là số nguyên tố
    
    for (int p = 2; p * p <= max_num; ++p) {
        if (is_prime[p]) {
            for (int i = p * p; i <= max_num; i += p) {
                is_prime[i] = false; // Đánh dấu các bội số của p
            }
        }
    }
    return is_prime;
}

int count_primes_in_range(int A, int B) {
    // Tạo danh sách số nguyên tố từ 1 đến B
    vector<bool> is_prime = sieve_of_eratosthenes(B);
    
    // Đếm số lượng số nguyên tố trong đoạn [A, B]
    int count = 0;
    for (int i = A; i <= B; ++i) {
        if (is_prime[i]) {
            count++;
        }
    }
    return count;
}

int main() {
    ifstream infile("CAU02.INP");
    ofstream outfile("CAU02.OUT");

    int A, B;
    infile >> A >> B; // Đọc A và B từ tệp

    // Đếm số lượng số nguyên tố
    int result = count_primes_in_range(A, B);

    // Ghi kết quả ra tệp
    outfile << result << endl;

    infile.close();
    outfile.close();

    return 0;
}
